File: hackerearth/practice/algorithms/dynamic-programming/introduction-to-dynamic-programming-1/samu-and-shopping/main.py

def func(a, n):
    memd={}         # memd reverses the order of be and id memd[be][id]

    def mins(ids, be):
        if(memd.has_key(be)):
            if(memd[be].has_key(ids)):
                return memd[be][ids]
        else:
            memd[be]={}
    
        if(ids==1): # id-1 is shirt. It says the box before current 'be' is shirt
            if(be==n):
                memd[be][ids]=min(a[be][2],a[be][3])
            else:
                memd[be][ids]=min(a[be][2]+mins(2, be+1), a[be][3]+mins(3, be+1))
    
        elif(ids==2): # id-2 is pant. It says the box before current 'be' is pant
            if(be==n):
                memd[be][ids]=min(a[be][1], a[be][3])
            else:
                memd[be][ids]=min(a[be][1]+mins(1, be+1), a[be][3]+mins(3, be+1))
    
        elif(ids==3): # id-1 is shoe. It says the box before current 'be' is shoe
            if(be==n):
                memd[be][ids]=min(a[be][1], a[be][2])
            else:
                memd[be][ids]=min(a[be][1]+mins(1, be+1), a[be][2]+mins(2, be+1))
        return memd[be][ids]

    return min(a[1][1]+mins(1, 2),a[1][2]+mins(2, 2),a[1][3]+mins(3, 2))

t=int(raw_input())
for i in xrange(t):
    a={}
    n=int(raw_input())
    for j in xrange(n):
        a[j+1]={}
        s=raw_input().split(' ')
        for k in xrange(3):
            a[j+1][k+1]=int(s[k])
    print func(a,n)

File: hackerearth/practice/algorithms/dynamic-programming/introduction-to-dynamic-programming-1/samu-and-shopping/main.cpp
#include<iostream>
#include<map>
#include<algorithm>

using namespace std;

long t,n;
map<long, map<long, long>> memd, mp;

long minm(long a, long b, long c){
    return min(min(a,b),c);
}

long mins(long id, long be){

    if(memd.find(be)!=memd.end()){
        if(memd[be].find(id)!=memd[be].end())
            return memd[be][id];
    }
    else{
        memd[be]=map<long, long>();
    }
    
    switch(id){
        case 1:
            if(be==n)
                memd[be][id]=min(mp[be][2], mp[be][3]);
            else
                memd[be][id]=min(mp[be][2]+mins(2, be+1), mp[be][3]+mins(3, be+1));
            break;

        case 2:
            if(be==n)
                memd[be][id]=min(mp[be][1], mp[be][3]);
            else
                memd[be][id]=min(mp[be][1]+mins(1, be+1), mp[be][3]+mins(3, be+1));
            break;

        case 3:
            if(be==n)
                memd[be][id]=min(mp[be][1], mp[be][2]);
            else
                memd[be][id]=min(mp[be][1]+mins(1, be+1), mp[be][2]+mins(2, be+1));
            break;

        return memd[be][id];

    }
}

long func(){
    if(n==1)
        return minm(mp[1][1], mp[1][2], mp[1][3]);
    return minm(mp[1][1]+mins(1, 2), mp[1][2]+mins(2, 2), mp[1][3]+mins(3, 2));
}

int main(){

    cin>>t;
    for(long i=0;i<t;i++){
        memd.clear();
        mp.clear();
        cin>>n;
        for(long j=0;j<n;j++){
            mp[j+1]=map<long, long>();
            for(long k=0;k<3;k++){
                mp[j+1][k+1]=0;
                cin>>mp[j+1][k+1];
            }
        }
        cout << func() << endl;
    }

    return 0;
}
